File: src/main/java/com/amazonaws/sample/cdk/discourse/CognitoNestedStack.java
package com.amazonaws.sample.cdk.discourse;

import software.amazon.awscdk.*;
import software.amazon.awscdk.services.cognito.*;
import software.constructs.Construct;

import java.util.List;

public class CognitoNestedStack extends NestedStack {

    private UserPool userPool;
    private UserPoolClient userPoolClient;

    public CognitoNestedStack(final Construct scope, final String id, final String domainName, final String cognitoAuthSubDomainName) {
        super(scope, id);
        createCognitoUserPool(id);
        createCognitoUserPoolDomain(cognitoAuthSubDomainName, id);
        createCognitoUserPoolClient(domainName, id);
    }

    public UserPool getUserPool() {
        return userPool;
    }

    public UserPoolClient getUserPoolClient() {
        return userPoolClient;
    }

    private void createCognitoUserPool(final String id) {
        userPool = UserPool.Builder.create(this, id + "UserPool")
                .removalPolicy(RemovalPolicy.DESTROY)
                .email(UserPoolEmail.withCognito())
                .signInCaseSensitive(false)
                .signInAliases(SignInAliases.builder().email(true).phone(false).username(false).build())
                .mfa(Mfa.OPTIONAL)
                .mfaSecondFactor(MfaSecondFactor.builder().otp(true).sms(true).build())
                .accountRecovery(AccountRecovery.EMAIL_AND_PHONE_WITHOUT_MFA)
                .autoVerify(AutoVerifiedAttrs.builder().email(true).build())
                .enableSmsRole(true)
                .passwordPolicy(PasswordPolicy.builder()
                        .minLength(8)
                        .requireDigits(true)
                        .requireSymbols(true)
                        .requireUppercase(true)
                        .requireLowercase(true)
                        .tempPasswordValidity(Duration.days(7))
                        .build())
                .selfSignUpEnabled(true)
                .standardAttributes(
                        StandardAttributes.builder()
                                .email(StandardAttribute.builder()
                                        .required(true)
                                        .mutable(true)
                                        .build())
                                .address(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .birthdate(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .familyName(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .gender(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .givenName(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .locale(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .middleName(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .fullname(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .nickname(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .phoneNumber(StandardAttribute.builder()
                                        .required(true)
                                        .mutable(true)
                                        .build())
                                .profilePicture(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .preferredUsername(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .profilePage(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .timezone(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .lastUpdateTime(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .website(StandardAttribute.builder()
                                        .mutable(true)
                                        .build())
                                .build())
                .build();
    }

    private void createCognitoUserPoolDomain(final String cognitoAuthSubDomainName, final String id) {
        UserPoolDomain.Builder.create(this, id + "UserPoolDomain")
                .userPool(userPool)
                .cognitoDomain(CognitoDomainOptions.builder()
                        .domainPrefix(cognitoAuthSubDomainName)
                        .build())
                .build();
    }

    private void createCognitoUserPoolClient(final String callbackDomain, final String id) {
        ClientAttributes clientReadAttributes = new ClientAttributes().withStandardAttributes(StandardAttributesMask.builder()
                .address(true)
                .birthdate(true)
                .email(true)
                .emailVerified(true)
                .familyName(true)
                .gender(true)
                .givenName(true)
                .locale(true)
                .middleName(true)
                .fullname(true)
                .nickname(true)
                .phoneNumber(true)
                .phoneNumberVerified(true)
                .profilePicture(true)
                .preferredUsername(true)
                .profilePage(true)
                .timezone(true)
                .lastUpdateTime(true)
                .website(true)
                .build());
        ClientAttributes clientWriteAttributes = new ClientAttributes().withStandardAttributes(StandardAttributesMask.builder()
                .address(true)
                .birthdate(true)
                .email(true)
                .familyName(true)
                .gender(true)
                .givenName(true)
                .locale(true)
                .middleName(true)
                .fullname(true)
                .nickname(true)
                .phoneNumber(true)
                .profilePicture(true)
                .preferredUsername(true)
                .profilePage(true)
                .timezone(true)
                .lastUpdateTime(true)
                .website(true)
                .build());
        userPoolClient = UserPoolClient.Builder.create(this, id + "UserPoolApp")
                .supportedIdentityProviders(List.of(
                        UserPoolClientIdentityProvider.COGNITO)
                )
                .oAuth(OAuthSettings.builder()
                        .callbackUrls(List.of("https://" + callbackDomain + "/auth/oidc/callback"))
                        .flows(OAuthFlows.builder()
                                .authorizationCodeGrant(true)
                                .build())
                        .scopes(List.of(OAuthScope.PHONE,
                                OAuthScope.EMAIL,
                                OAuthScope.OPENID,
                                OAuthScope.COGNITO_ADMIN,
                                OAuthScope.PROFILE))
                        .build())
                .refreshTokenValidity(Duration.days(30))
                .idTokenValidity(Duration.minutes(60))
                .enableTokenRevocation(true)
                .generateSecret(true)
                .preventUserExistenceErrors(true)
                .userPool(userPool)
                .readAttributes(clientReadAttributes)
                .writeAttributes(clientWriteAttributes)
                .authFlows(AuthFlow.builder()
                        .adminUserPassword(false)
                        .userPassword(false)
                        .userSrp(false)
                        .custom(false)
                        .build())
                .build();
    }
}

File: lib/cognito-nested-stack.ts
import {Construct} from 'constructs';
import {Duration, NestedStack, NestedStackProps, RemovalPolicy} from "aws-cdk-lib";
import {
    AccountRecovery,
    ClientAttributes,
    Mfa,
    OAuthScope,
    UserPool,
    UserPoolClient,
    UserPoolClientIdentityProvider,
    UserPoolDomain,
    UserPoolEmail
} from "aws-cdk-lib/aws-cognito";

export class CognitoNestedStack extends NestedStack {
    userPool: UserPool;
    userPoolClient: UserPoolClient;

    constructor(scope: Construct, id: string, props?: NestedStackProps) {
        super(scope, id, props);

        this.userPool = new UserPool(this, id + "UserPool", {
            removalPolicy: RemovalPolicy.DESTROY,
            email: UserPoolEmail.withCognito(),
            signInCaseSensitive: false,
            signInAliases: {
                email: true,
                username: false,
                phone: false
            },
            mfa: Mfa.OPTIONAL,
            mfaSecondFactor: {
                otp: true,
                sms: true
            },
            accountRecovery: AccountRecovery.EMAIL_AND_PHONE_WITHOUT_MFA,
            autoVerify: {
                email: true
            },
            enableSmsRole: true,
            passwordPolicy: {
                minLength: 8,
                requireDigits: true,
                requireSymbols: true,
                requireUppercase: true,
                requireLowercase: true,
                tempPasswordValidity: Duration.days(7),
            },
            selfSignUpEnabled: true,
            standardAttributes: {
                email: {required: true, mutable: true},
                address: {mutable: true},
                birthdate: {mutable: true},
                familyName: {mutable: true},
                gender: {mutable: true},
                givenName: {mutable: true},
                locale: {mutable: true},
                middleName: {mutable: true},
                fullname: {mutable: true},
                nickname: {mutable: true},
                phoneNumber: {required: true, mutable: true},
                profilePicture: {mutable: true},
                preferredUsername: {mutable: true},
                profilePage: {mutable: true},
                timezone: {mutable: true},
                lastUpdateTime: {mutable: true},
                website: {mutable: true},
            }
        });

        new UserPoolDomain(this, id + "UserPoolDomain", {
            userPool: this.userPool,
            cognitoDomain: {
                domainPrefix: process.env.CDK_DEPLOY_DISCOURSE_COGNITO_AUTH_SUB_DOMAIN_NAME || ''
            }
        });

        const clientReadAttributes = (new ClientAttributes()).withStandardAttributes({
            address: true,
            birthdate: true,
            email: true,
            emailVerified: true,
            familyName: true,
            gender: true,
            givenName: true,
            locale: true,
            middleName: true,
            fullname: true,
            nickname: true,
            phoneNumber: true,
            phoneNumberVerified: true,
            profilePicture: true,
            preferredUsername: true,
            profilePage: true,
            timezone: true,
            lastUpdateTime: true,
            website: true
        });
        const clientWriteAttributes = (new ClientAttributes()).withStandardAttributes({
            address: true,
            birthdate: true,
            email: true,
            familyName: true,
            gender: true,
            givenName: true,
            locale: true,
            middleName: true,
            fullname: true,
            nickname: true,
            phoneNumber: true,
            profilePicture: true,
            preferredUsername: true,
            profilePage: true,
            timezone: true,
            lastUpdateTime: true,
            website: true
        });
        this.userPoolClient = new UserPoolClient(this, id + "UserPoolApp", {
            supportedIdentityProviders: [
                UserPoolClientIdentityProvider.COGNITO
            ],
            oAuth: {
                callbackUrls: [
                    'https://' + (process.env.CDK_DEPLOY_DISCOURSE_DOMAIN_NAME || '') + '/auth/oidc/callback'
                ],
                flows: {
                    authorizationCodeGrant: true
                },
                scopes: [
                    OAuthScope.PHONE,
                    OAuthScope.EMAIL,
                    OAuthScope.OPENID,
                    OAuthScope.COGNITO_ADMIN,
                    OAuthScope.PROFILE
                ]
            },
            refreshTokenValidity: Duration.days(30),
            idTokenValidity: Duration.minutes(60),
            enableTokenRevocation: true,
            generateSecret: true,
            preventUserExistenceErrors: true,
            userPool: this.userPool,
            readAttributes: clientReadAttributes,
            writeAttributes: clientWriteAttributes,
            authFlows: {
                adminUserPassword: false,
                userPassword: false,
                userSrp: false,
                custom: false
            }
        });
    }
}
